general_bounds.c

```diff
@@ -1,45 +1,52 @@
-#include "include/common.h"
-#include <stdint.h>
-#include <stdio.h>
-#include <stdlib.h>
-#include <string.h>
-
 /***
  * This program is a fusion of ``bounds.c``
  * and ``set_bounds.c``.
  * CHERI limits the range of addresses that may be dereferenced
  * and reading the value outside the range will always give
  * an "In-address space security exception". However, if we try to
  * extend the bounderies of a capable pointer by setting them explicitly,
- * the program fails on "riscv64" but not on "morello".
+ * the program fails on "riscv64".
  ***/
 
+#include <cheri/cheric.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "include/common.h"
+
 int main()
 {
 	int32_t array[12] = {0};
 	int32_t *typed_array = &array;
 	u_int32_t bounds = 48;
 
-	printf("Explicitly setting the bounds outside the range ");
-	if (strcmp(ARCH, "Morello") == 0)
-	{
-		printf("does not cause any failure on %s. It automatically clears them.\n", ARCH);
-	}
-	else
-	{
-		printf("causes the following exception: ");
-	}
-
+#if defined(__aarch64__)
+	// Do a dereference
 	uint64_t length = cheri_getlength(typed_array);
 	for (uint32_t counter = 0; counter <= (length / sizeof(int32_t)) + 11; counter++)
 	{
 		inspect_pointer(typed_array + counter);
 		// Read value to crash
 		if (counter == 12)
 		{
-			puts("--> On any architecture dereferencing outside the range"
-				 " causes the following exception:");
+			printf("--> On any architecture dereferencing outside the range"
```

Comment:
I think we can now change this from "on any architecture" to "On Morello" (or should it be "on AArch64"? can we differentiate Morello from other Arm architectures here?).

Reply:
Dereferencing outside the range, as far as I know, it is forbidden on any architecture. Again, dereferencing, not explicitly setting the bounds.

Reply:
Yes, but now this code is guarded by `#defined` it can't be called on anything that isn't AArch64. [Which is why I'm wondering if we can actually restrict it to Morello, since not every AArch64 platform is Morello.]

Reply:
But we did the check on "Aarch64", you are quite right. Well, we wanted an example working in different ways on "riscv64" and "Morello". If we put an ``||`` to allow the same on "riscv6" then the whole point of the general example will be missed, i.e. on "riscv64" it would crash before reaching the other branch.

Reply:
> I think we can now change this from "on any architecture" to "On Morello" (or should it be "on AArch64"? can we differentiate Morello from other Arm architectures here?).
"Morello" is ARMv8, I think that ``clang`` macros allow only to get that "v8".

Reply:
> If we put an || to allow the same on "riscv64" then the whole point of the general example will be missed

That's not what I'm suggesting.

My point is that if you do:

```c
#if defined(aarch64)
// at this point this code can only be on aarch 64
#endif
```

then the `#ifdef` code implicitly excludes other platforms so the "on any architecture" warning is a bit misleading, because that code won't be compiled on anything other than AArch64.

> "Morello" is ARMv8, I think that clang macros allow only to get that "v8".

Then I guess we want something like `#if defined(__aarch64__) and armversion = v8` to capture "Morello"?

Reply:
Right. I'll try to capture that "v8", then.